### DISCUSSION TITLE
Using CIPDriver with a PowerFlex drive

### DISCUSSION THREAD

**Discussion by rnoguera2** [ottowayi/pycomm3#195] [Q&A] [answered]
Hi,

I am trying to use CIPDriver to read a parameter from a connected PowerFlex 525 VFD drive.
I am quite sure the path is correct, as **drive._list_identity()** is giving me the correct drive information.
For the other parameters I am sticking to the example in the documentation:
https://docs.pycomm3.dev/en/latest/examples/generic_messaging_examples.html#accessing-drive-parameters

But I am getting the same output all the time:
##################################
pf525_param, b'', None, Service not supported
Press any key to continue . . .
##################################

I tryed with different services and parameters, with the same result.
As a verification, I have been able to read the same parameter, from the same VFD from a PLC using the built-in MSG instruction...

Is anyone using current version succesfully with a VFD?

Thanks in advance,

**Comment by rnoguera2** [comment, marked as answer]
Hi,

I got it working, by removing a couple of parameters from the example code:

def read_pf525_parameter():
    drive_path = '172.22.186.70'
    
    
    with CIPDriver(drive_path) as drive:
        param = drive.generic_message(
            service=Services.get_attribute_single,
            class_code=b'\x93',
            instance=41,  # Parameter 41 = Accel Time
            attribute=b'\x09',
            data_type=INT,
            connected=False,
            #unconnected_send=True,
            #route_path=True,
            name='pf525_param'
        )
        print(param)

read_pf525_parameter()

**Reply by ottowayi** [reply]
Sorry I missed this post, yeah it appears those settings are the issue.  The example uses a routed path, so they are needed in that case.  I will update the docs to remove those.  The `unconnected_send` arg will be going away all together, since it is only used for unconnected but routed requests it will become automatic in the next version.

**Comment by ClearSimpleAndPlain** [comment]
I'm new to GitHub.  I wish I had found these discussions earlier.  After trying lots and much research I was able to get it to work myself.  I used class x0F (parameter) instead of x93.  I wonder what the difference is.


```
from pycomm3 import (CIPDriver, Services, ClassCode, CommonClassAttributes, INT)
#read a vfd parameter

with CIPDriver("10.0.0.1") as driver: #VFD address
    resp = driver.generic_message(
               service=Services.get_attribute_single, #get attribute single
               class_code=ClassCode.parameter,        #parameter class
               instance=41, 			      #41dec = Accel Time
               attribute=0x01, 			      #value attribute
               data_type=INT,
               connected=False 			      #don't do an open forward (faster for one read?)
               )
    print(resp)
```
<edited: took forever to get the formatting down.  got it now>

**Reply by ottowayi** [reply]
what model of VFD do you have?  I think the class codes may be different between the different families, like 525 vs 40.  Usually you can find the class codes in knowledge base docs that have examples on doing it thru a MSG instruction in the PLC.

**Reply by ClearSimpleAndPlain** [reply]
It was a PowerFlex525. Class x93 attr 9 also works for the 525.

**Comment by rnoguera2** [comment]
Hi,

This values depend on the VFD family.
You can find them in the product manuals.

If you were communicating with the VFD from a ControlLogix PLC you would be using the same values. This makes easy to identify them on some of the messaging examples.

Regards,
